src/qibo/backends/__init__.py

```diff
@@ -78,39 +79,95 @@ class GlobalBackend(NumpyBackend):
-            cls._instance = construct_backend(backend, platform=platform)
+            backend = construct_backend(backend_env, platform=platform)
         else:
             # Create backend according to default order
             for kwargs in cls._default_order:
                 try:
-                    cls._instance = construct_backend(**kwargs)
+                    backend = construct_backend(**kwargs)
                     break
-                except (ImportError, MissingBackend):
+                except (ModuleNotFoundError, ImportError):
                     pass
 
-        if cls._instance is None:  # pragma: no cover
+        if backend is None:  # pragma: no cover
             raise_error(RuntimeError, "No backends available.")
+        return backend
 
-        log.info(f"Using {cls._instance} backend on {cls._instance.device}")
-        return cls._instance
+    @classmethod
+    def set_backend(cls, backend, **kwargs):
+        cls._backend = construct_backend(backend, **kwargs)
+        log.info(f"Using {cls._backend} backend on {cls._backend.device}")
 
     @classmethod
-    def set_backend(cls, backend, **kwargs):  # pragma: no cover
-        if (
-            cls._instance is None
-            or cls._instance.name != backend
-            or cls._instance.platform != kwargs.get("platform")
-        ):
-            cls._instance = construct_backend(backend, **kwargs)
-        log.info(f"Using {cls._instance} backend on {cls._instance.device}")
+    def transpiler(cls):
+        """Get the current transpiler. If no transpiler is set, it will create one."""
+        if cls._transpiler is not None:
+            return cls._transpiler
+
+        cls._transpiler = cls._default_transpiler()
+        return cls._transpiler
+
+    @classmethod
+    def set_transpiler(cls, transpiler):
+        cls._transpiler = transpiler
+        # TODO: check if transpiler is valid on the backend
+
+    @classmethod
+    def _default_transpiler(cls):
+        import networkx as nx
+
+        from qibo.transpiler.optimizer import Preprocessing
+        from qibo.transpiler.pipeline import Passes
+        from qibo.transpiler.placer import Trivial
+        from qibo.transpiler.router import Sabre
+        from qibo.transpiler.unroller import NativeGates, Unroller
+
+        qubits: list[str] = list(cls._backend.qubits)  # list of qubit names
+        natives: list[str] = cls._backend.natives
+        connectivity_edges: list[tuple[str, str]] = (
+            cls._backend.connectivity
+        )  # list of edges
+        if qubits and natives and connectivity:
+
+            # code is needed in unroll.py to convert str -> enum
+            natives_enum = NativeGates.from_gatelist(natives)
+
+            connectivity = nx.Graph()
+
+            # q{i} naming
+            node_mapping = {qubits[i]: i for i in range(len(qubits))}
+            for e in connectivity_edges:
+                connectivity.add_edge(node_mapping[e[0]], node_mapping[e[1]])
```

Comment:
A graph can be initialized from its edges in a single batch, you don't need to add them one by one.
https://networkx.org/documentation/stable/reference/classes/generated/networkx.Graph.__init__.html

Reply:
- `qubits` contains `str` hardware qubit names.
- `connectivity` cannot have `str` as nodes.
- `connectivity_edges` has `tuple[str, str]`.

We need to add edges by iterating over `connectivity_edges` and converting the qubit names to integers based on the mapping.

Reply:
Or we can add like this.

```py
connectivity = nx.Graph()

node_mapping = {qubits[i]: i for i in range(len(qubits))}
connectivity.add_edges_from((node_mapping[e[0]], node_mapping[e[1]]) for e in connectivity_edges)
```

Reply:
Yes, I was proposing exactly something as in your last comment (the bottom one seems the cleanest).

Moreover, the presence of `node_mapping` is just a performance optimization, since you're just building a reverse index (you already have the forward map `i -> qubits[i]`, but inverting it for specific values would require continuous searches, so `node_mapping` is essentially doing the scan just once and caching).

A couple of minor improvements are:
- you could decide for a simpler name for `node_mapping`, e.g. `idxs`, since it's only locally relevant (pure aesthetics)
- there is a built-in function doing exactly what you're doing with `range(len())`, i.e. `enumerate()`
	```py
	node_mapping = {q: i for i, q in enumerate(qubits)}
	```
	(you only need to swap the tuple, and dict comprehension is definitely the simplest way of doing that)